src/graph/LookupExecutor.cpp

```diff
@@ -0,0 +1,499 @@
+    if (ids.size() == 1) {
+        index_ = ids[0];
+        return status;
+    }
+    /**
+     * Returns the first index if no optimization is required
+     */
+    if (skipOptimize_) {
+        index_ = ids[0];
+        return status;
+    }
+    auto *mc = ectx()->getMetaClient();
+
+    /**
+     * std::map<column_hint_count, IndexID>
+     */
+    std::map<int32_t, IndexID> indexHint;
+    for (auto& indexId : ids) {
+        auto itemStatus = (isEdge_) ? mc->getEdgeIndexFromCache(spaceId_, indexId)
+                                    : mc->getTagIndexFromCache(spaceId_, indexId);
+        if (!itemStatus.ok()) {
+            return itemStatus.status();
+        }
+        const auto& index = itemStatus.value();
+        int32_t hintCount = 0;
+        for (const auto& field : index->get_fields()) {
+            auto it = std::find_if(filters_.begin(), filters_.end(),
+                                   [field](const auto &rel) {
+                                       return field.get_name() == rel.first &&
+                                              rel.second == RelationalExpression::EQ;
+                                   });
+            if (it == filters_.end()) {
+                break;
+            }
+            ++hintCount;
+        }
+        /**
+         * return the index of max hint count.
+         */
+        indexHint[hintCount] = indexId;
+    }
+    index_ = indexHint.rbegin()->second;
+    return status;
+}
+
+std::vector<IndexID>
+LookupExecutor::findValidIndex(const std::vector<std::shared_ptr<nebula::cpp2::IndexItem>>& ids) {
+    std::vector<IndexID> indexes;
+    for (auto& index : ids) {
+        bool indexHint = true;
+        if (index->get_schema_name() != *from_) {
+            continue;
+        }
+        const auto& fields = index->get_fields();
+        for (auto& col : filters_) {
+            auto it = std::find_if(fields.begin(), fields.end(),
+                                   [col](const auto &field) {
+                                       return field.get_name() == col.first;
+                                   });
+
+            // If the property name not find in index's field
+            // we need break this loop then check next index.
+            if (it == fields.end()) {
+                indexHint = false;
+                break;
+            }
+        }
+        if (!indexHint) {
+            continue;
+        }
+        indexes.emplace_back(index->get_index_id());
+    }
+    return indexes;
+}
+
+void LookupExecutor::finishExecution(std::unique_ptr<RowSetWriter> rsWriter) {
+    auto outputs = std::make_unique<InterimResult>(std::move(returnCols_));
+    if (rsWriter != nullptr) {
+        outputs->setInterim(std::move(rsWriter));
+    }
+
+    if (onResult_) {
+        onResult_(std::move(outputs));
+    } else {
+        resp_ = std::make_unique<cpp2::ExecutionResponse>();
+        auto colNames = outputs->getColNames();
+        resp_->set_column_names(std::move(colNames));
```

Comment:
If no onResult_,  we could skip the extra encode/decode,  you could recheck the logic in GoExecutor.